### DISCUSSION TITLE
Loading tables from other schemas for Oracle datasources

### DISCUSSION THREAD

**Discussion by sebastian-jansson** [Budibase/budibase#10206] [Help]
Been trying to get this to work without any luck.

After checking the source for the oracle integration I can see that my workaround on setting another schema using alter session after my user logs on will not work.

**Comment by melohagan** [comment]
Hey @sebastian-jansson 

SQL Views are currently not supported in Budibase: https://github.com/Budibase/budibase/issues/6216

**Reply by sebastian-jansson** [reply]
Apologies @melohagan , a typo on my end, I was referring to schemas. Thanks for the quick reply!

Since I posted this I've read through the oracle.ts integration source code and found a potential solution. 

However, I'm having issues with getting the dev env to run under Ubuntu (22.04) so i can't test it yet.

The tab never finishes loading, no errors in the terminal apart from some minor 
Will create another topic on that issue if I can't get it running on my own soon.

**Comment by sebastian-jansson** [comment]
I've finally managed to set up the development environment successfully! 
The culprit was my ad blocker, _uBlock Origin_, which of course is installed on all my browsers :)  Lesson learned for sure.

### Resolving the schema limitation
To resolve my issue of not being able to load in tables from other schemas, I made a quick adjustment to the **COLUMNS_SQL** variable within `packages/server/src/integrations/oracle.ts` by simply modifying the query to fetch data from `all_tables` instead of `user_tables`. 

By setting the all_tables.OWNER column to either the username or schema name (if provided) I'm assuming normal functionality when not specifying a schema would be affected.

Initially, I attempted to add the schema name as a new field in the edit data source view. However, it seems it wasn't as straightforward as I'd hoped. I'm guessing there might be some more locations where changes have to be made.

###  Workaround
As I wanted to focus on building a prototype of my app rather than troubleshooting further, I temporarily hard coded my schema name instead, which works perfectly for now.

Fetching tables now works as I intended and I see all the tables from the schema  _(For which my user has select permissions on)_

### Feature viability
@melohagan Do you think this would be a viable PR to merge in if we make the necessary adjustments so the schema field would show up by the data source?

### Modified SQL
 ```
 SELECT
    tabs.table_name,
    cols.column_name,
    cols.data_type,
    cols.data_default,
    cols.column_id,
    cons.constraint_name,
    cons.constraint_type,
    cons.r_constraint_name,
    cons.search_condition
  FROM
    all_tables tabs
  JOIN 
    all_tab_columns cols
    ON tabs.table_name = cols.table_name 
  LEFT JOIN 
    all_cons_columns col_cons
    ON cols.column_name = col_cons.column_name
    AND cols.table_name = col_cons.table_name
  LEFT JOIN 
    all_constraints cons
    ON col_cons.constraint_name = cons.constraint_name
    AND cons.table_name = cols.table_name
  WHERE
     (cons.status = 'ENABLED' OR cons.status IS NULL)
  AND tabs.owner = '%schema_name%'
```

### Diff
![image](https://user-images.githubusercontent.com/54795430/230483112-4536de23-9a02-4abf-a59a-dee879590ebd.png)

**Reply by melohagan** [reply]
Hey @sebastian-jansson 

Thank you very much for your detailed resolution. I think @mike12345567 would be the best person to ask here whether using `all_tables` over `user_tables` is suitable.

**Reply by sebastian-jansson** [reply]
@mike12345567 any thoughts on this?

**Reply by mike12345567** [reply]
Hi @sebastian-jansson - I think this could work, in `oracle.ts` we would need to update the `SCHEMA` structure to make it so that when configuring the datasource it takes a `schema` property, e.g.
```
schema: {
  type: DatasourceFieldType.STRING,
}
```

I don't think we can require this as lots of users are using the system as it is today - and won't have a schema value configured. Instead, I think we could use your new `COLUMNS_SQL` if a schema property has been set on the datasource, otherwise fallback to the current version which utilises the `user_tables`.

Also - great work on debugging this!
